DatadogInternal/Sources/DataStore/DataStore.swift

```diff
@@ -0,0 +1,83 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+
+/// Represents the version of data format stored for a key in the data store.
+///
+/// As values are persisted, the format of data may change between different SDK versions. To prevent decoding errors,
+/// callers can utilize `DataStoreKeyVersion` to explicitly specify the version of serialized `Data` before attempting deserialization.
+public typealias DataStoreKeyVersion = UInt16
+
+/// The default version of data stored for keys (equals `0`).
+public let dataStoreDefaultKeyVersion: DataStoreKeyVersion = 0
+
+/// Possible results of retrieving a value from the data store.
+public enum DataStoreValueResult {
+    /// The value was found and serialized using the format defined by the specified version.
+    case value(Data, DataStoreKeyVersion)
+    /// There was no value associated with the requested key.
+    case noValue
+    /// The value could not be read due to an underlying error.
+    /// This may represent an error with the file format or an I/O exception that occurred during reading.
+    case error(Error)
+
+    /// Retrieves the data value associated with the result, if it matches the expected version.
+    ///
+    /// - Parameter expectedVersion: The version expected for the retrieved data.
+    /// - Returns: The data value if the version matches the expected version; otherwise, nil.
+    public func data(expectedVersion: DataStoreKeyVersion = dataStoreDefaultKeyVersion) -> Data? {
+        guard case .value(let data, let storedVersion) = self, storedVersion == expectedVersion else {
+            return nil
+        }
+        return data
+    }
```

Comment:
Alternatively we can change the semantic to make it always return `Data` (if found) with performing an optional check on `expectedVersion` (if provided) as in following code. I have no strong opinion.
```swift
public func data(expectedVersion: DataStoreKeyVersion? = nil) -> Data? {
    guard case .value(let data, let storedVersion) = self else {
        return nil
    }
    if let expectedVersion == expectedVersion, storedVersion != expectedVersion {
        return nil
    }

    return data
}
```

Reply:
not sure if version expectation should be built in.

I would assume that the consumer of the api will handle different model version migration

Reply:
The `version` is now made a pure convenience. It is fully hidden in the basic usage:
```swift
store.setValue(data, forKey: "key")
store.value(forKey: "key") { result in
   _ = result.data()
}
```

If only versioning problem occurs, we can leverage it:
```swift
store.setValue(data_in_new_version, forKey: "key", version: 2)
store.value(forKey: "key") { result in
   _ = result.data(expectedVersion: 2)
}
```

I think this is the best balance between having this support and keeping a clean API. Requiring the caller to handle versioning will cause unnecessary repetitive code infection across different places.

Reply:
> Requiring the caller to handle versioning will cause unnecessary repetitive code infection across different places.

Also, if `version` is not encoded for every data by upfront, we won't be able to ever support migration from "no version" to "version: 2". The decoding failure will become ambiguous and (if not filtered somehow) will flood telemetry with false-positives.